### PR TITLE
MODLink System (+ NWTLMM)

### PR DESCRIPTION
**Pull request by Fikou**
## About The Pull Request
A pact made with @Kapu1178 
Small changes you should not care about:
RD MODsuit outfit (admin only) no longer has a beret that blocks the activation of the suit
The beret used by death squad officers no longer is blocked from being put on a hat stabilizer module
Admins can now Shear matrices of objects in Modify Transform
Multitool buffers have been a little refactored to use a setter proc that saves them from causing hard dels
Cooler stuff:
A revival and remake of [Nobody Wants To Learn Matrix Math](https://github.com/tgstation/tgstation/pull/59103), this time with additional tooling for quick matrix calculations.
image
The MODLink system, available through every MODsuit and MODLink scryers (a neck item obtainable from advanced modsuit research or charliestation)
Let's you make a holographic call with any other MODLink user, where you can chat in realtime and see what's up with em
image image

## Why It's Good For The Game
Adds a fun way for the crew to communicate with each other that can be done in real-time with relative privacy compared to radio.

## Changelog
:cl: Fikou, Armhulen, Sheets (+rep for Mothblocks and Potato)
fix: RD MODsuit outfit (admin only) no longer has a beret that blocks the activation of the suit
fix: The beret used by death squad officers no longer is blocked from being put on a hat stabilizer module
admin: Admins can now Shear matrices of objects in Modify Transform
admin: Admins now have access to Test Matrices in the VV dropdown, an all-in-one tool for editing transforms.
add: MODLink system, available through scryers (from RnD and Charlie Station) and through MODsuits. Lets you call people with holographs!
/:cl:

### CODE REVIEW

**Comment by Paxilmaniac** [comment]
modlink 2023

**Comment by Donglesplonge** [comment]
cool ranch 👍

**Comment by Vincent983** [comment]
can you call/get calls while having an inactive modsuit on your back?

**Comment by Fikou** [comment]
> can you call/get calls while having an inactive modsuit on your back?

nope it uses cell power

File: code/modules/mod/mod_control.dm [hunk 1]
```diff
@@ -587,6 +591,9 @@
 	QDEL_LIST_ASSOC_VAL(old_module.pinned_to)
 	old_module.mod = null
 
+/obj/item/mod/control/proc/get_wearer()
```
Comment (MrMelbert) [inline code/modules/mod/mod_control.dm:594]: Add a doc to this explaining it's for use in callbacks and not general use

File: code/modules/mod/mod_link.dm [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+	var/mob/living/user = mod_link.get_user_callback.Invoke()
+	playsound(mod_link.get_other().holder, 'sound/machines/terminal_processing.ogg', 50, vary = TRUE, frequency = -1)
+	LAZYREMOVE(mod_link.holder.update_on_z, mod_link.visual)
+	mod_link.holder.lose_hearing_sensitivity(REF(mod_link))
+	mod_link.holder.UnregisterSignal(user, list(COMSIG_CARBON_APPLY_OVERLAY, COMSIG_CARBON_REMOVE_OVERLAY, COMSIG_ATOM_DIR_CHANGE))
+	QDEL_NULL(mod_link.visual)
+
+/proc/on_user_set_dir_generic(datum/mod_link/mod_link, newdir)
+	var/atom/other_visual = mod_link.get_other().visual
+	if(newdir == null)
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:41]:
```suggestion
	if(isnull(newdir))
```
Though if newdir should never be null, consider using `ASSERT`

File: code/modules/admin/view_variables/nobody_wants_to_learn_matrix_math.dm [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+/datum/nobody_wants_to_learn_matrix_math
+	var/atom/target
+	var/matrix/testing_matrix
+
+/datum/nobody_wants_to_learn_matrix_math/New(atom/target)
+	src.target = target
+	testing_matrix = matrix(target.transform)
+
+/datum/nobody_wants_to_learn_matrix_math/Destroy(force, ...)
+	QDEL_NULL(testing_matrix)
```
Comment (MrMelbert) [inline code/modules/admin/view_variables/nobody_wants_to_learn_matrix_math.dm:19]: are matrices objects? do you need to qdel them?

File: code/modules/mod/mod_link.dm [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+		other_visual.pixel_x = -14
+		other_visual.layer = BELOW_MOB_LAYER
+		SET_PLANE_IMPLICIT(other_visual, GAME_PLANE_FOV_HIDDEN)
+		new_transform.Shear(-0.5, 0)
+		new_transform.Scale(0.65, 1)
+	other_visual.transform = new_transform
+
+/obj/item/mod/control
+	var/datum/mod_link/mod_link
+	var/starting_frequency
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:73]: Autodocs

File: code/modules/mod/mod_link.dm [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+
+/obj/item/mod/control/Initialize(mapload, datum/mod_theme/new_theme, new_skin, obj/item/mod/core/new_core)
+	. = ..()
+	mod_link = new(src, starting_frequency, CALLBACK(src, PROC_REF(get_wearer)), CALLBACK(src, PROC_REF(can_call)), CALLBACK(src, PROC_REF(make_link_visual)), CALLBACK(src, PROC_REF(get_link_visual)), CALLBACK(src, PROC_REF(delete_link_visual)))
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:77]: Multiline this? Hard to read with all these callbacks

File: code/modules/mod/mod_link.dm [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+	var/tool_frequency = null
+	if(istype(tool.buffer, /datum/mod_link))
+		var/datum/mod_link/buffer_link = tool.buffer
+		tool_frequency = buffer_link.frequency
+	if(!tool_frequency && mod_link.frequency)
+		tool.set_buffer(mod_link)
+	else if(tool_frequency && !mod_link.frequency)
+		mod_link.frequency = tool_frequency
+	else if(tool_frequency && mod_link.frequency)
+		var/response = tgui_alert(user, "Would you like to copy or imprint the frequency?", "MODLink Frequency", list("Copy", "Imprint"))
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:91]: Should put some sanity checks after this alert. Stop people input stalling to set frequency whenever / wherever they want.

File: code/modules/mod/mod_link.dm [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+	else
+		call_link(user, mod_link)
+
+/obj/item/clothing/neck/link_scryer/proc/get_user()
+	var/mob/living/carbon/user = loc
+	return istype(user) && user.wear_neck == src ? user : null
+
+/obj/item/clothing/neck/link_scryer/proc/can_call()
+	var/mob/living/user = loc
+	return istype(user) && cell.charge && user.stat < DEAD
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:214]: Docs docs docs docs docs

File: code/modules/mod/mod_link.dm [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+	var/id = ""
+	var/atom/movable/holder
+	var/atom/movable/visual
+	var/list/visual_overlays = list()
+	var/datum/mod_link_call/link_call
+	var/datum/callback/get_user_callback
+	var/datum/callback/can_call_callback
+	var/datum/callback/make_visual_callback
+	var/datum/callback/get_visual_callback
+	var/datum/callback/delete_visual_callback
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:268]: Docs docs docs docs docs (+ a class level doc would be nice)

File: code/modules/mod/mod_link.dm [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+
+/datum/mod_link/New(atom/holder, frequency, datum/callback/get_user_callback, datum/callback/can_call_callback, datum/callback/make_visual_callback, datum/callback/get_visual_callback, datum/callback/delete_visual_callback)
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:270]: Multiline this

File: code/modules/mod/mod_link.dm [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+/datum/mod_link/proc/end_call()
+	QDEL_NULL(link_call)
+
+/datum/mod_link/proc/on_holder_delete(atom/source)
+	SIGNAL_HANDLER
+	qdel(src)
+
+/datum/mod_link_call
+	var/datum/mod_link/caller
+	var/datum/mod_link/receiver
```
Comment (MrMelbert) [inline code/modules/mod/mod_link.dm:337]: Small class level doc would be appreciated for this one as well

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/admin/view_variables/nobody_wants_to_learn_matrix_math.dm (hunk 3 above)
Comment (Fikou) [reply code/modules/admin/view_variables/nobody_wants_to_learn_matrix_math.dm:19]: i think so

**Comment by Vincent983** [comment]
could you add some way to rename the modlink scryers without a labeler?

**Review by Wallemations — APPROVED** [review]
Sprites look fine

**Comment by DATA-xPUNGED** [comment]
This really should have a way to be used with an inactive modsuit. It's just weird that even with a modsuit always on (and with the slowdown) you still pretty much need to have one on your neck

**Comment by DaydreamIQ** [comment]
Hey so there doesn't seem to be a way to recharge the scryers once a cell is put in, and the cell seems to be impossible to remove at that

**Comment by Vincent983** [comment]
@DaydreamIQ did you try right-clicking the scryer? it took out the cell when I did it

**Comment by LSommers** [comment]
Having seen the scryer/MODLinl in action, that shit is _very cool_ and frankly makes sense as a progression of holopad tech ingame. Not something that can really be OP either- if someone jumps the dude with the giant floating holographic voice call screen in front of their face and gets caught, they deserve it for having the situational awareness of a fucking Rhododendron.

**Comment by TheVekter** [comment]
Balls

**Comment by Vincent983** [comment]
The MODlink is cool as hell now that I've tested it, but could you allow lathes to also print it? It being only in the exosuit fabricator will likely limit how many people will use it, combined with needing advanced modsuits to be researched in order to print it.

**Comment by Fikou** [comment]
> The MODlink is cool as hell now that I've tested it, but could you allow lathes to also print it? It being only in the exosuit fabricator will likely limit how many people will use it, combined with needing advanced modsuits to be researched in order to print it.

idk what lathe would fit it

**Comment by Vincent983** [comment]
Could stick it in the science lathe in addition to the exofab

File: code/modules/research/techweb/all_nodes.dm [hunk 11]
```diff
@@ -1648,6 +1648,7 @@
 		"mod_clamp",
 		"mod_drill",
 		"mod_orebag",
+		"modlink_scryer"
```
Comment (san7890) [inline code/modules/research/techweb/all_nodes.dm:1651]:
```suggestion
		"modlink_scryer",
```

**Review by san7890 — CHANGES_REQUESTED** [review]
i'm going to assume direct control on this PR

**Comment by kittysmooch** [comment]
these bad boys dont work with TTS devices fwiw

**Pull request merged**